### PR TITLE
feat!: Upgrade cosmos-sdk to v0.46 

### PR DESCRIPTION
**Pull request by yihuang**
## Description

prepare for implementing cosmos-sdk feemarket and tx prioritization

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: app/utils.go [hunk 1]
```diff
@@ -50,11 +54,15 @@ var DefaultConsensusParams = &abci.ConsensusParams{
 
 // Setup initializes a new EthermintApp. A Nop logger is set in EthermintApp.
 func Setup(isCheckTx bool, patchGenesis func(*EthermintApp, simapp.GenesisState) simapp.GenesisState) *EthermintApp {
-	db := dbm.NewMemDB()
+	return SetupWithDB(isCheckTx, patchGenesis, dbm.NewMemDB())
```
Comment (fedekunze) [inline app/utils.go:57]: I feel this improvement can be merged in another PR?

File: app/app.go [hunk 2]
```diff
@@ -566,39 +568,65 @@ func NewEthermintApp(
...
+	app.SetAnteHandler(anteHandler)
+}
 
-	if loadLatest {
-		if err := app.LoadLatestVersion(); err != nil {
-			tmos.Exit(err.Error())
-		}
+func (app *EthermintApp) setPostHandler() {
+	postHandler, err := posthandler.NewPostHandler(
+		posthandler.HandlerOptions{},
```
Comment (fedekunze) [inline app/app.go:623]: we should refund leftover gas here instead on the EVM

File: app/ante/ante.go [hunk 3]
```diff
@@ -23,7 +23,19 @@ const (
 // Ethereum or SDK transaction to an internal ante handler for performing
 // transaction-level processing (e.g. fee payment, signature verification) before
 // being passed onto it's respective handler.
-func NewAnteHandler(options HandlerOptions) sdk.AnteHandler {
+func NewAnteHandler(options HandlerOptions) (sdk.AnteHandler, error) {
+	if options.AccountKeeper == nil {
```
[the remark this answers is not in the document]
Comment (yihuang) [reply app/ante/ante.go:27]: It seems the method is removed

File: app/app.go (hunk 2 above)
Comment (yihuang) [reply app/app.go:623]: `posthandler` is run for the whole tx rather than individual tx msg, so maybe not a good place to do gas refund?

File: app/ante/ante.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yihuang) [reply app/ante/ante.go:27]: I changed to use the `Validate` method?

File: app/utils.go (hunk 1 above)
Comment (yihuang) [reply app/utils.go:57]: extracted to https://github.com/evmos/ethermint/pull/1171

**Comment by yihuang** [comment]
tests all passed.

**Comment by faddat** [comment]
I'll apply this branch to the chain in a pr.

Nice work!

**Comment by faddat** [comment]
Changed import paths:

https://github.com/yihuang/ethermint/pull/110

File: app/app.go (hunk 2 above)
Comment (fedekunze) [reply app/app.go:622]: I think it's a cleaner solution

**Comment by yihuang** [comment]
wait for: https://github.com/cosmos/ibc-go/pull/1653

**Comment by fedekunze** [comment]
@yihuang we can remove ibc-go from the `go.mod` to unblock this and readd it once we they update it

**Comment by yihuang** [comment]
> @yihuang we can remove ibc-go from the `go.mod` to unblock this and readd it once we they update it

but we do use ibc-go in ante handler and `app.go`, it won't build?

**Comment by fedekunze** [comment]
We also need to update the `third_party/` proto directory with the newest files btw

File: app/app.go (hunk 2 above)
Comment (fedekunze) [reply app/app.go:622]: Let's create an issue to implement the Post Handler for Ethereum transactions

**Review by fedekunze — APPROVED** [review]
ACK

**Comment by fedekunze** [comment]
Follow ups:

- Implement the Post Handler for Ethereum transactions
- update `third_party/` proto directory with the newest files
- Tx Prioritization #1194
